File: test/react-native/features/fixtures/reactnative/module/ConfigFileReader.kt
package com.reactnative.module

import android.content.Context
import org.json.JSONObject
import java.io.File
import java.io.IOException

const val CONFIG_FILE_TIMEOUT = 5000

class ConfigFileReader {

    fun getMazeRunnerAddress(context: Context): String {
        val externalFilesDir = context.getExternalFilesDir(null)
        val configFile = File(externalFilesDir, "fixture_config.json")
        var mazeAddress: String? = null
        Log.i("Bugsnag", "Attempting to read Maze Runner address from config file ${configFile.path}")

        // Poll for the fixture config file
        val pollEnd = System.currentTimeMillis() + CONFIG_FILE_TIMEOUT
        while (System.currentTimeMillis() < pollEnd) {
            if (configFile.exists()) {
                val fileContents = configFile.readText()
                val fixtureConfig = runCatching { JSONObject(fileContents) }.getOrNull()
                mazeAddress = getStringSafely(fixtureConfig, "maze_address")
                if (!mazeAddress.isNullOrBlank()) {
                    Log.i("Bugsnag", "Maze Runner address set from config file: $mazeAddress")
                    break
                }
            }

            Thread.sleep(250)
        }
        if (mazeAddress.isNullOrBlank()) {
            Log.i("Bugsnag", "Failed to read Maze Runner address from config file, reverting to legacy address")
            mazeAddress = "bs-local.com:9339"
        }
        return mazeAddress
    }

    private fun getStringSafely(jsonObject: JSONObject?, key: String): String {
        return jsonObject?.optString(key) ?: ""
    }

}

File: test/react-native/features/fixtures/scenario-launcher/lib/ConfigFileReader.js
import { Platform } from 'react-native'
import { Dirs, FileSystem } from 'react-native-file-access'

const TIMEOUT = 60000

// Used when no config file can be read at all. This is only ever correct when maze
// runner is reachable on the device itself (i.e. a tunnelled/local run).
const FALLBACK_ADDRESS = 'localhost:9339'

const delay = ms => new Promise(resolve => setTimeout(resolve, ms))

const getCandidatePaths = () => {
  if (Platform.OS === 'ios') {
    return [`${Dirs.DocumentDir}/fixture_config.json`]
  }

  return [
    '/sdcard/Android/data/com.reactnative/files/fixture_config.json',
    '/storage/emulated/0/Android/data/com.reactnative/files/fixture_config.json',
    `${Dirs.SDCardDir}/Android/data/com.reactnative/files/fixture_config.json`,
    '/data/local/tmp/fixture_config.json',
    `${Dirs.SDCardDir}/fixture_config.json`,
    `${Dirs.DocumentDir}/fixture_config.json`,
    `${Dirs.CacheDir}/fixture_config.json`
  ]
}

// Reads the maze runner address that maze runner pushes to the device. Waits up to
// `timeout` ms for the file to appear; pass 0 to read whatever is on disk right now.
const getMazeRunnerAddress = async (timeout = TIMEOUT) => {
  const candidatePaths = getCandidatePaths()
  const startTime = Date.now()

  // poll for the config file to exist
  while (true) {
    for (const path of candidatePaths) {
      try {
        const configFileExists = await FileSystem.exists(path)

        if (configFileExists) {
          const configFile = await FileSystem.readFile(path)
          console.error(`[Bugsnag ConfigFileReader] found config file at '${path}'. contents: ${configFile}`)
          const config = JSON.parse(configFile)
          if (config && config.maze_address) {
            return `${config.maze_address}`
          }
        }
      } catch (err) {
        // Continue searching other candidate paths
      }
    }

    if (Date.now() - startTime >= timeout) break

    await delay(500)
  }

  console.error(`[Bugsnag ConfigFileReader] no config file found in candidate paths, falling back to '${FALLBACK_ADDRESS}'`)
  return FALLBACK_ADDRESS
}

module.exports.getMazeRunnerAddress = getMazeRunnerAddress
module.exports.FALLBACK_ADDRESS = FALLBACK_ADDRESS